### ISSUE TITLE
require('ttypescript') alongside require('typescript')

### ISSUE THREAD

**Issue by cspotcode** [cevek/ttypescript#19]
Is it supported to programmatically use an unpatched typescript alongside the patched version from ttypescript?  I see that if I `require('ttypescript')` then `require('typescript')` gets patched, whereas ideally ttypescript creates a separate instance of the compiler.

I think this would do the trick, but I'm not sure if you will accept a PR for it.  Are there reasons to keep your current implementation?

```
"use strict";
import * as fs from 'fs';
import * as Path from 'path';
import {patchCreateProgram} from 'ttypescript/lib/patchCreateProgram';

// Create a new copy of the typescript library
function createTypescriptInstance() {
    const typescriptLibPath = require.resolve('typescript/lib/typescript');
    const typescriptLibSourceText = fs.readFileSync(typescriptLibPath, 'utf8');
    const __filename = typescriptLibPath;
    const __dirname = Path.dirname(__filename);
    const exports = {}, module = {exports};
    eval(typescriptLibSourceText);
    return module.exports as any;
}

module.exports = patchCreateProgram(createTypescriptInstance());
```

**Comment by cevek** [comment]
I think it can be done more simpler:
```ts
import * as ts from 'typescript';
import {patchCreateProgram} from './patchCreateProgram';
const patchedTs = {...ts};
export = patchCreateProgram(patchedTs);
```

This issue related to #18
We need to find a compromise

**Comment by cspotcode** [comment]
I don't think `{...ts}` will work because all of typescript's functions are bound to the original `ts` object.  For example, when `transpileModule` internally calls `createProgram` it will not invoke the patched `createProgram` because we have not mutated the original `ts` object.

With `eval` we can construct a new, isolated copy of the library.

**Comment by cevek** [comment]
All createProgram calls are called with exported `ts` - `ts.createProgram()`

**Comment by cspotcode** [comment]
`ts` is bound to the original `exports` object.  Doing `{...ts}` creates a *new* object, but calls to `ts.createProgram()` are referring to the *original* `exports` object.  They will not invoke the replacement implementation of `createProgram`.

You can verify this yourself by running the code example you posted.

```
import * as ts from 'typescript';
import {patchCreateProgram} from 'ttypescript/lib/patchCreateProgram';
const patchedTs = {...ts};
patchedTs.createProgram = function () {throw 'patched implementation invoked'};
console.log(patchedTs.transpileModule('const foo: any;', {}).outputText);
// Result: exception is *not* thrown because patched implementation is ignored.  This method of patching fails
```

```
tsc code-example.ts ; node code-example.js
```

Am I missing something?

**Comment by cspotcode** [comment]
@cevek I just read #18.

What if we allow `require('ttypescript/clone')`?  "clone" will be a separate instance of the typescript library; it will not patch `require('typescript')`.  People who need to patch `require('typescript')` can use require('ttypescript');`

Is that a good compromise?

**Comment by cevek** [comment]
I've made refactoring, now ttypescript makes copy of original typescript

**Comment by cspotcode** [comment]
Cool, thanks!

**Comment by cevek** [comment]
Fixed in ttypescript@1.5.2
